### DISCUSSION TITLE
Security Concern: Use of Weak Hashing Algorithm in node_modules/pg/lib/crypto/utils-webcrypto.js

### DISCUSSION THREAD

**Discussion by furkantopal** [mikro-orm/mikro-orm#5910] [General]
### Describe the bug

Hello Mikro-ORM team,

Our application recently underwent a security analysis, and it flagged a potential security issue related to the use of a weak hashing algorithm in one of your dependencies. Specifically, the concern points to node_modules/pg/lib/crypto/utils-webcrypto.js.

The analysis warns that the application utilizes a hashing algorithm that does not meet current security standards. Weak hashing algorithms can lead to vulnerabilities, such as the leaking of sensitive information or credentials. The primary issue is the potential use of an outdated or insecure hash function that could undermine the overall security of the application.

To address this issue, we recommend replacing the weak hash function with a stronger one. For example, you can use the SHA-256 algorithm.

Thank you for your attention to this matter, and we appreciate your ongoing efforts to maintain and improve Mikro-ORM.

Best regards

### Reproduction

Observe the usage of the hashing algorithm in node_modules/pg/lib/crypto/utils-webcrypto.js.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

5.9.2

### Node.js version

20.12

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> 5.9.2

First of all, v5 is not supported anymore.

> Specifically, the concern points to node_modules/pg/lib/crypto/utils-webcrypto.js.

Which is our dependency, why are you reporting this here instead of at [their end](https://github.com/brianc/node-postgres)? We can't do anything about code we don't maintain.

I can imagine the code in question will look differently now, since you use old version of the ORM and the dependency on `pg` is pinned. If you update to latest v6, you will get latest version of `pg` too. 

Upgrading should be your first step either way.

**Comment by pehota** [comment]
The md5 algorithm is being use here: https://github.com/mikro-orm/mikro-orm/blob/22b7e97040792d7b56676daab99de69a5ae58b6c/packages/core/src/utils/Utils.ts#L959
It's a security alert in one of the apps I am working on too.

**Reply by B4nan** [reply]
Well, that is used only to compute random strings for file names in cache. How is that a security concern?

**Reply by B4nan** [reply]
And also for randomization of too long index names, that's the other use case, its not used anywhere else, I don't see any way this could be abused.

**Reply by B4nan** [reply]
I really dont mind changing this to `sha256` if it will resolve some security alert about our code (unlike what the OP is about), but it could as well be just `(Math.random() + 1).toString(36).substring(7)` or something similar, this is not a password hashing function or something like that, we don't need "perfectly random values" here.

**Reply by pehota** [reply]
Thank you for the detailed clarification, @B4nan !
To give you some context and help you decide if you want to update that part of the code, the report we're getting from our security monitoring app is: `Public tools can be used to exploit this standard vulnerability type`.
I am not sure the tool is clever enough to crawl your code and make sense of the md5 usage.
Now that I know what you're using the the hashed string for, I think I will lower the severity of the error in our tool and leave it up to you to decide if you want to change this or not.
Thanks!

**Reply by B4nan** [reply]
And the alert is specifically about using `md5`?

**Reply by pehota** [reply]
Yes. Here's a screenshot from the report.
<img width="216" alt="image" src="https://github.com/user-attachments/assets/c85ca223-6e39-4d98-a4a1-7a413a9c26da">

**Reply by B4nan** [reply]
I'll try to just swap this to `sha256` later today, it should all work the same I guess, it's not really important what is there, it's only used during the init phase, never on runtime.

**Reply by B4nan** [reply]
Btw this won't change anything about the use of `md5` inside `pg` dependency 🙃 That one is still there even in latest version.

https://github.com/brianc/node-postgres/blob/master/packages/pg/lib/crypto/utils-webcrypto.js#L35

**Reply by B4nan** [reply]
And the only real alternative to that package is also having that in the codebase 🤡 

https://github.com/porsager/postgres/blob/f58cd4f3affd3e8ce8f53e42799672d86cd2c70b/cjs/src/connection.js#L997

This really feels like a false alert you should just ignore IMO. But I'll resolve the use on our end for the sake of better future, heh.

**Reply by pehota** [reply]
Thanks a lot, @B4nan !

**Reply by B4nan** [reply]
Changing this feels actually a tiny bit breaking, since it would result in different computed names for long FKs and indexes, so extra migrations.

**Reply by furkantopal** [reply]
If this change offers no real benefit and only introduces breaking changes and unnecessary hurdles, we can choose not to proceed with it if you prefer. @B4nan

**Comment by furkantopal** [comment]
Hi @B4nan, I also opened an issue on **pg** end.
Many thanks!
